Comments: Add real data to the comment management

Copons opened the pull request:
Closes #14488 

With this PR we add some real data to the comment management!

## Changes

- Move the list of comments in a `CommentList` component that contains both the list and the section navigation, and manages all the various actions (approve, like, trash, etc.).
This is against what we decided to do previously (cc @dmsnell ), but my reasoning is that makes sense to have these actions in a single place, if they are to be performed by both `CommentDetail` and `CommentNavigation`.

- Obtain comments data from the Redux state via the `getSiteComments` selector.
Right now, we get _all_comments, and then we filter them in the list. This is not the correct behaviour, but I couldn't get the query component to output the filtered list fast enough. Also, changing the list in the props would require more unnecessary work to handle comments both in state and in props.

- Wire the Redux comments (in `this.props`) into the internal state (`this.state`).
This way we can test the behaviour without performing destructive operations. After reloading the page, all the comments will be there again.

- The actions modify the internal state and move comments around in the various status lists.
For example an Approved comment can be marked as Spam: it will then disappear from the Approved list and appear in the Spam list.

- Display the correct actions for the various statuses.
Also adds a "Delete Permanently" button on Spam and Trash (it's indistinguishable from the normal Trash button, but I didn't want to do style work in this PR).

- Fix typechecking errors and replaced the term "pending" with "unapproved".

- Replace the multiple checks in `CommentDetailActions` with action lists.

- Replace `commentIsApproved`, `commentIsSpam` and `commentIsTrash` with the simpler `commentStatus`.

cc @Automattic/lannister

dmsnell commented on client/blocks/comment-detail/index.jsx:
```diff
@@ -188,32 +184,27 @@ export class CommentDetail extends Component {
+	postUrl: get( ownProps, 'post.link' ),
+	repliedToComment: ownProps.replied, // TODO: not available in the current data structure
+	siteId: ownProps.siteId,
+} );
```
Something I hope to do soon is get a comment object defined somewhere and write a mapping function from this API call into that object. @rodrigoi I'll make sure you are part of it too. That way, in the end we should be able to safely use nesting of properties where appropriate. For example, it might eventually make more sense to use `comment.isSpam` or `const { isSpam, isLiked } = comment` instead of `commentIsSpam`. Further, I'd like to challenge us to think outside the box on some of these properties and dig deep into the essence of what a comment _is_. For example, a comment _cannot_ be both `spam` _and_ `trash`, so why does our data object allow that invalid representation?

Copons replied:
No good reason I'm afraid. Those properties are from my first iteration of the comment, before I realized that all those information are stored in `comment.status`.
I was about to remove them altogether, but I didn't want to increase the size of this PR even more.
It might make sense to remove them after a first run of reviews, though.

gwwar replied:
> For example, a comment cannot be both spam and trash, so why does our data object allow that invalid representation?

That makes sense to remove the redundant booleans and stick with a status enum. Though we'll likely want to keep`isLiked`. We can have an approved comment, that is not liked for example.

> I was about to remove them altogether, but I didn't want to increase the size of this PR even more.

I wouldn't mind if you refactored a bit here. We don't need to be as cautious when the components are not in use anywhere in prod.

dmsnell replied:
no worries - I wasn't suggesting doing the work here now. by all means, let's build on what we have before changing to what we don't

gwwar commented on client/blocks/comment-detail/comment-detail-actions.jsx:
```diff
@@ -12,24 +12,27 @@ export const CommentDetailActions = ( {
-					: translate( 'Like' )
-			}</span>
-		</a>
+		{ ! commentIsSpam && ! commentIsTrash &&
```
How would folks feel about passing through an explicit list of actions to render? I think we could avoid having to pass through so many booleans then, if we pass through status, isLiked, and an action list.

Where actions might look like:
```
 [ 'like', 'approve', 'spam' 'trash' ]
```
or
```
 [ 'approve', 'delete' ]
```

gwwar replied:
Feel free to think about this for future PRs

dmsnell replied:
in the notifications we had to be aware of the fact that some comments don't have all those actions - we have a tri-state device where the presence of the key means it's _likeable_ and the value is whether _it's liked_.

https://github.com/Automattic/notifications-panel/blob/e27c2ce64f086994f9769a79a7bc056309e81d86/src/templates/actions.jsx#L55

could have similar needs here

gwwar commented:
Awesome @Copons feel free to 🚢  this! Folks can start to chat about the behavior already. For example, do we need some sort of transition when we click on `approve`, or would a notice be enough to signal what happened?

gwwar commented:
Oh another small thing I noticed is that we're escaping the comment content. @dmsnell are these safe to inject directly, or should we be parsing for text content?

![screen shot 2017-05-26 at 10 12 46 am](https://cloud.githubusercontent.com/assets/1270189/26505023/e695f396-41fb-11e7-9be9-5854399085ea.png)

Copons commented:
@gwwar I was thinking of adding a delayed fade/slide out, but didn't want to add any styling here.

dmsnell commented:
> @dmsnell are these safe to inject directly, or should we be parsing for text content?

this should be safe and should be parsed on the server. if we detect an anomaly it should be secured in the API-side (think `wp_kses()` and `strip_tags()` and friends)

gwwar commented:
Nice, I wasn't sure if we had some corner cases for Jetpack site.

Copons commented:
@dmsnell @gwwar 

I'm gonna ship this now, but just FYI these are the last changes:

- Replaced the multiple checks in `CommentDetailActions` with action lists as proposed by @gwwar.

- Replaced `commentIsApproved`, `commentIsSpam` and `commentIsTrash` with the simpler `commentStatus`, as proposed by @dmsnell.

dmsnell commented:
@copons we should review those `translate()` calls at some point. Some of theme should have a `context` attached, such as `translate( 'Trash', { context: 'verb' } )` because the word could be used differently in English depending on the context (could also be a noun, for instance)
